runtime/signals.c

```diff
@@ -248,7 +255,8 @@ void caml_execute_signal(int signal_number, int in_signal_handler)
     caml_sigmask_hook(SIG_SETMASK, &sigs, NULL);
   }
 #endif
-  if (Is_exception_result(res)) caml_raise(Extract_exception(res));
+  if (Is_exception_result(res))
+    caml_raise_in_async_callback(Extract_exception(res));
```

Comment:
Do we know for sure that we are called from an async callback here, or is the idea that it's harmless to call this even if we aren't? (Then maybe a comment could clarify that.)

It's clear that we are in a postponed callback if we were called from `caml_process_sending_signals`, but what about the calls protected by `caml_try_leave_blocking_section_hook()` in `signals_{byt,nat}.c`?

Reply:
I'll clarify the notion of "asynchronous callback": I mean any callback triggered through `caml_check_urgent_gc`.

However, you are right that `caml_raise_in_async_callback` can be called by memprof even if not going through `caml_check_urgent_gc`. I'll add a comment for saying that calling this function instead of `caml_raise` does not hurt (apart from the performance penalty caused by an extra needless call to `caml_check_urgent_gc`).